Generic GnssDevice

torbacz opened the pull request:
<!--- In the TITLE (↑↑↑↑ above ↑↑↑↑ **NOT HERE**) provide a general, short summary of your changes -->
<!--- Please DO NOT use references to other PR's or issues -->

## Description
<!--- Describe your changes in detail -->
<!--- Bulleted list. Full sentences. Ending with a dot. -->
- New project for all GPS devices across NF project.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If this **fixes** OR **closes** OR  **resolves** an open issue, please link to the issue there using the template bellow (mind the pattern to link there as all issues are tracked in the Home repository) -->
<!--- **JUST** replace NNNNN with the issue number -->
- As the build with project reference failes to build on pipeline I decieded to split it into 2 separate PRs and use package reference in GPS module.


## Types of changes
<!--- What types of changes does this PR introduce? Put an `x` in all the boxes that apply: -->
- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- PLEASE PLEASE PLEASE don't tick all of them just because -->
- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
	- Introduced a categorization system for GPS and GNSS devices, enhancing organization and interaction.
	- Implemented structured dependency management for both GPS and GNSS devices, ensuring consistent builds.
	- Established versioning and release management configurations for both GPS and GNSS projects to improve deployment processes.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Ellerbach commented on devices/GpsDevice/NMEA0183Parser.cs:
```diff
@@ -0,0 +1,90 @@
+    /// <summary>
+    /// Provides methods for parsing NMEA0183 data from a GPS device.
+    /// </summary>
+    public static class NMEA0183Parser
```
Question: shall we follow our naming rules and use Nmea0183Parser instead?

Ellerbach commented on devices/GpsDevice/NMEA0183Parser.cs:
```diff
@@ -0,0 +1,90 @@
+        /// </summary>
+        /// <param name="inputData">The valid GNGLL data string starting with $GNGLL.</param>
+        /// <returns>The parsed GngllData object.</returns>
+        public static GngllData ParaseGngll(string inputData)
```
as all the NMEA string are fixed, can you create a kind of generic parser?
Something that can give one of the GngllData or GngsaData file?
I would see a generalized based class names "NmeaData" which contains a field with the signature.
As an example, it could look like:

```csharp
public interface NmeaData
{
    public string Id {get; }
    public NmeaData Parse(string inputData);
}

public class GngsaData: NmeaData
{
    public override string Id = "$GNGSA";
   public NmeaData Parse(string inputData);
    // Rest of the class
}

// And in the parser

public static class NMEA0183Parser
{
    public NmeaParser[] MneaDatas {get; protected set; }

      // In the constructor, just add all the NmeaData derivated class that are supported
     //  MneaDatas = new MneaData[] { GngsaData }; // add all supported

    public NmeaData Parse(string inputData)
    {
        var data = inputData.Split(',');
        for each (string id in NmeaDatas)
        {
            if (id.Id == data[0])
            {
                return id.Parse(inputData);
            }
        }
    }
}
```

Advantage: you can add later on any parser you'd like!

Ellerbach commented on devices/GpsDevice/category.txt:
```diff
@@ -0,0 +1,2 @@
+gps
```
not sure we have those categories yet, please check the project that builds automatically all the readmes! If needed, can you add those into the categoies? (sorry, a bit in a rush to check into details)

Ellerbach reviewed (commented):
I love it! See one of my comment to make it extendible easily!

networkfusion commented:
Would it not  be better to call this the GnssDevice base project to better represent that it can be used for Galilao etc.?

Ellerbach commented:
> Would it not be better to call this the GnssDevice base project to better represent that it can be used for Galilao etc.?

Yes, good catch btw! That's indeed a great idea. We can later on add the elements. Reminds me I've been working on GNSS for the SIM Com modems. Changes are still in a branch (which is btw ready for PR): https://github.com/nanoframework/nanoFramework.IoT.Device/tree/add-atmodem-gnss/devices/AtModem

So most likely elements to scaange there!

networkfusion commented:
> > Would it not be better to call this the GnssDevice base project to better represent that it can be used for Galilao etc.?
> 
> Yes, good catch btw! That's indeed a great idea. We can later on add the elements. Reminds me I've been working on GNSS for the SIM Com modems. Changes are still in a branch (which is btw ready for PR): https://github.com/nanoframework/nanoFramework.IoT.Device/tree/add-atmodem-gnss/devices/AtModem
> 
> So most likely elements to scaange there!

I also want to draw attention to the https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/device/geolocation?view=net-maui-8.0&tabs=windows namespace...

Ellerbach commented:
> Maybe we can hijack the API and rename the namespace to nanoFramework...
> If the API is sound, of course, which I haven't looked into yet. 

So I had a look at all this and I don't think it's relevant for a device. Basically what's there is a service to get a geolocalisation. Which needs to use a device, totally hidden in that case. Here, it's kind of the opposite. BUT, we can try to reuse the same names for the Localisation and few others liks the events.

@torbacz I don't have much time but I think I can help a bit with this PR so that we can use it both for your other PR and the SIMCom modem. So that all classes, base class, enums, namespaces are aligned. Let me know if that's ok for you, so I can work on a this PR as well!

torbacz commented:
@Ellerbach Sure, feel free to commit :)

torbacz commented on devices/GnssDevice/Nmea0183Parser.cs:
```diff
@@ -0,0 +1,114 @@
+        /// </summary>
+        /// <param name="inputData">The valid GNGLL data string starting with $GNGLL.</param>
+        /// <returns>The parsed GngllData object.</returns>
+        public static GngllData ParaseGngll(string inputData)
```
@Ellerbach Should we remove it from parser?

Ellerbach replied:
yes, yes, again, sorry, miss to push a commit yesterday, please check now, should be all good.

torbacz commented on devices/GnssDevice/Nmea0183Parser.cs:
```diff
@@ -0,0 +1,114 @@
+        /// <summary>
+        /// Gets or sets a list of NMEA0183 data objects.
+        /// </summary>
+        public static INmeaData[] MneaDatas { get; set; }
```
How this will be populated?

Ellerbach replied:
I just realized I missed a push :-( All is here now

Ellerbach commented:
@torbacz I think I fixed all the core elements, including the tests (you can't pass enums directly, so need to be casted). See the tests which are now running with nano.
Things to be added:
* Extract a timestamp for the commands an add it to the GeoLocalisation
* Depending on the commands, populate more elements (like altitude but also the precision)
* Add a README
* Add few more very used commands?

torbacz commented:
@Ellerbach 
I've pushed some changes:
- Changed how MneaDatas are handled (Hashtable instead of collection)
- Changed Nmea0183Parser.Parse return type to object (this interface won't work in my opinion as not all data will have the same fields - this way it's much safer in my opinion. BTW I really miss generic, this is a perfect case for them :( )
- Added some "Debug" error handling 
- Some docs fixes and changed visibility of methods

Let me know If you are ok with such version.

As for TODOs, I would go with minimal version right now. Then we can add more/adjust existing commands.

Ellerbach commented:
@torbacz 
> Changed Nmea0183Parser.Parse return type to object (this interface won't work in my opinion as not all data will have the same fields

as you've simplified the interface and anyway only classes with this interface can be processed, better return the strong type

for the rest, added functions to add or remove from the collection, so it's a bit safer

and added time and date management

There is a test to fix (you created it originally) and then we can indeed go with this first version

torbacz commented:
@Ellerbach 
Test fixed, so I think we are ready to merge this version?

Ellerbach reviewed (approved):
Let's go for this first version. I'll most likely add more elements in another PR. I'm happy with the current implementation to start with! Thanks again @torbacz for the great collaboration.
